### ISSUE TITLE
Cannot obtain Guest OS call graph

### ISSUE THREAD

**Issue by ustiugov** [firecracker-microvm/firecracker#1153]
To profile the whole processing pipeline from the host to the guest and back for my toy TCP server, I am trying to obtain the call graph of the **guest and host** from the host to generate a per-syscall/kernel-function-call/guest-kernel-function-call breakdown of the latency per packet. 
I would like to get the following info (copied from [here](https://lwn.net/Articles/378778/)):
```
            38770.00 20.4% __ticket_spin_lock        [guest.kernel.kallsyms]
            22560.00 11.9% ftrace_likely_update      [kernel.kallsyms]
             9208.00  4.8% __lock_acquire            [kernel.kallsyms]
```

However, perf produces no samples in perf.data for the guest OS. 
```
sudo perf kvm --guest --guestmount=/tmp/guestmount record -a sleep 10
sudo perf kvm --guest --guestmount=/tmp/guestmount record -a -e cpu-clock:GH -g -o perf.data
^C[ perf record: Woken up 10 times to write data ]
[ perf record: Captured and wrote 3.720 MB perf.data (25457 samples) ]

sudo perf kvm --guest --guestmount=/tmp/guestmount report -i perf.data --stdio
# To display the perf.data header info, please use --header/--header-only options.
#
#
# Total Lost Samples: 0
#
# Samples: 25K of event 'cpu-clock:GH'
# Event count (approx.): 6364250000
# Children      Self  Command  Shared Object  Symbol
# ........  ........  .......  .............  ......
#


```
I recompiled the guest kernel with all the appropriate parameters. I am able to get the stack for the host (no guest kallsyms can be found in the output):
```
sudo perf kvm --host --guest --guestmount=/tmp/guestmount record -a -F 99 -g -o perf.data
sudo perf kvm --host --guest --guestmount=/tmp/guestmount report -i perf.data --stdio

# Samples: 40K of event 'cycles:ppp'
# Event count (approx.): 20557903667
#
# Overhead  Command          Shared Object            Symbol                                                                                                    
# ........  ...............  .......................  ..........................................................................................................
#
    38.37%  fc_vcpu0         [kernel.kallsyms]        [k] vmx_vcpu_run
     7.03%  fc_vcpu0         [kernel.kallsyms]        [k] kvm_lapic_find_highest_irr
     5.21%  fc_vcpu0         [kernel.kallsyms]        [k] kvm_arch_vcpu_runnable
     4.19%  fc_vcpu0         [kernel.kallsyms]        [k] __apic_update_ppr
     4.06%  fc_vcpu0         [kernel.kallsyms]        [k] ktime_get
     2.20%  fc_vcpu0         [kernel.kallsyms]        [k] kvm_vcpu_check_block
```
and from inside the guest:
```
perf record -F 99 -p 2936 -g sleep 10

# Samples: 218  of event 'cpu-clock'
# Event count (approx.): 2202020180
#
# Children      Self  Command          Shared Object        Symbol                            
# ........  ........  ...............  ...................  ..................................
#
    43.58%     5.96%  tcp_server  ld-musl-x86_64.so.1  [.] __clock_gettime
            |          
            |--40.37%--__clock_gettime
            |          |          
            |          |--36.70%--__vdso_clock_gettime
            |          |          |          
            |          |          |--15.60%--0x7ffc6e0d9991

```
But unfortunately not both host and guest call graph combined...
I would really appreciate any advice or suggestions on how I could fix or troubleshoot this problem.

**Comment by andreeaflorescu** [comment]
@ustiugov sorry for the late reply. We haven't played with ```perf kvm``` so far. Do you still have issues in getting the call graph for both guest and host? If that's the case I can try to reproduce this issue and let you know what I find out.

**Comment by ustiugov** [comment]
@andreeaflorescu Thank you for the response. Yes, the issue persists. 
TL;DR: `perf kvm --guest --host record ...` launched from the host shows only the host call graph but not the guest.

Host: Ubuntu 18, 4.15.0-52-generic
Guest: 4.18 (from firecracker-demo)

Please let me know if you need any further information.

**Comment by andreeaflorescu** [comment]
@ustiugov I think there might be some steps missing from your example. This [article](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/virtualization_tuning_and_optimization_guide/sect-virtualization_tuning_optimization_guide-monitoring_tools-perf_kvm) has a nice explanation of additional steps that are required to get the data you are looking for. I am specifically referring to Section 2.1. Wouldn't that be a required step to get data from the guest?

I am running ```sudo perf kvm --guest --host  --output perf.data``` and it looks like there's a lot of data recorded. I cannot analyse the data because there are some missing symbols:
```
Failed to open /tmp/perf-385.map, continuing without symbols
Failed to open /tmp/perf-10546.map, continuing without symbols
Failed to open /tmp/perf-5742.map, continuing without symbols
Failed to open /tmp/perf-5758.map, continuing without symbols
Failed to open /tmp/perf-10506.map, continuing without symbols
Failed to open /tmp/perf-10578.map, continuing without symbols
no symbols found in /bin/kmod, maybe install a debug package?
Failed to open /tmp/perf-10413.map, continuing without symbols
Failed to open /tmp/perf-10015.map, continuing without symbols
Failed to open /tmp/perf-5751.map, continuing without symbols
Failed to open /tmp/perf-5774.map, continuing without symbols
Failed to open /tmp/perf-17764.map, continuing without symbols
Failed to open /tmp/perf-14640.map, continuing without symbols
Failed to open /tmp/perf-10496.map, continuing without symbols
Failed to open /tmp/perf-10531.map, continuing without symbols
Failed to open /tmp/perf-29113.map, continuing without symbols
no symbols found in /usr/sbin/acpid, maybe install a debug package?
Failed to open /tmp/perf-11781.map, continuing without symbols
Failed to open /tmp/perf-11753.map, continuing without symbols
Failed to open /tmp/perf-4983.map, continuing without symbols
Failed to open /tmp/perf-5731.map, continuing without symbols
no symbols found in /bin/date, maybe install a debug package?
```

I am doing a wild guess and assume that the missing symbols are in the guest and you need to follow Section 2.1 from the linked article to get them.

**Comment by ustiugov** [comment]
if you do `perf kvm report`, do you see any [g] records? My problem is that I don't see any guest records. 
If instead of --guest and --host you only the former, so you see any samples recorded?

**Comment by ustiugov** [comment]
@andreeaflorescu Sorry, I forgot to put the name tag. Please see the comment above.

**Comment by andreeaflorescu** [comment]
@ustiugov I don't know what you mean with [g] records, I am most likely doing something wrong myself :))

**Comment by ustiugov** [comment]
If you type `sudo perf kvm --host --guest --guestmount=/tmp/guestmount report -i perf.data --stdio`, you should see something like this (however, you may not see the function names because you are missing symbols, probably just hex):

```
# Overhead  Command          Shared Object            Symbol                                                                                                    
# ........  ...............  .......................  ..........................................................................................................
#
    38.37%  fc_vcpu0         [kernel.kallsyms]        [k] vmx_vcpu_run
     7.03%  fc_vcpu0         [kernel.kallsyms]        [k] kvm_lapic_find_highest_irr
     5.21%  fc_vcpu0         [kernel.kallsyms]        [k] kvm_arch_vcpu_runnable
     4.19%  fc_vcpu0         [kernel.kallsyms]        [k] __apic_update_ppr
```

`[k]` means that this record belongs to the host kernel. If you grep the output, you should notice that the output does not contain any [g] records that correspond to Guest OS kernel. Please refer to this [doc](https://perf.wiki.kernel.org/index.php/Tutorial#Sample_analysis_with_perf_report).

**Comment by andreeaflorescu** [comment]
@ustiugov I checked and I see no output from the guest. Did you mounted the root file system of the guest on the host before running the record command?

**Comment by ustiugov** [comment]
@andreeaflorescu that's exactly my problem. 
I did mount the filesystem. To mount it, you can type:
```
mkdir -p /tmp/guestmount/YOUR_VM_PID
sshfs -o allow_other,direct_io root@169.254.0.1:/ /tmp/guestmount/YOUR_VM_PID -o IdentityFile=../firecracker-demo/xenial.rootfs.id_rsa
```

This would allow perf to access guest kernel's and guest kernel modules' symbols. In my case, the output stays empty.

**Comment by andreeaflorescu** [comment]
@ustiugov I am not very sure that this issue is related to Firecracker. Are you able to run these tests with another VMM? I would suggest trying out with Qemu to see if the problem you're seeing is related to the vmm.

**Comment by ustiugov** [comment]
@andreeaflorescu looks like the issue persists with QEMU too but I cannot find any bug report... Can it be true that we are the first to encounter this problem on a modern kernel? difficult to believe

**Comment by andreeaflorescu** [comment]
@ustiugov you can try asking on the [KVM email list](https://www.linux-kvm.org/page/Lists,_IRC#Mailing_Lists), maybe someone has a better suggestion on how to fix this problem.

I'll close this issue as it is not directly related to Firecracker.

**Comment by serban300** [comment]
@ustiugov I did some testing and this issue seems to reproduce only if the host kernel is less than 4.17. 

Could you try a version of kernel >= 4.17 and see if it works as expected ?

**Comment by ustiugov** [comment]
@serban300 today I tried with host kernel v4.18 but I still cannot get the guest call graph. Could you please post the commands you used to invoke `perf` and a sample of perf output?

**Comment by ustiugov** [comment]
I could answer my own request :) I managed to collect guest records:
```
     0.02%  :5175    [guest.kernel.kallsyms]  [g] do_splice_to
     0.02%  :5175    [guest.kernel.kallsyms]  [g] ipipe_prep
     0.01%  :5175    [unknown]                [u] 0x000055b5c4508bd0
     0.01%  :5175    [unknown]                [u] 0x00007ffd059ea981
     0.01%  :5175    [unknown]                [u] 0x00007f3cd3f91a6c
     0.01%  :5175    [guest.kernel.kallsyms]  [g] ext4_xattr_set_entry
     0.01%  :5175    [guest.kernel.kallsyms]  [g] propagate_umount
     0.00%  :5175    [guest.kernel.kallsyms]  [g] arch_optimize_kprobes
     0.00%  :5175    [guest.kernel.kallsyms]  [g] ovl_inode_update
     0.00%  :5175    [guest.kernel.kallsyms]  [g] ext4_remount
     0.00%  :5175    [guest.kernel.kallsyms]  [g] ext4_file_open
     0.00%  :5175    [guest.kernel.kallsyms]  [g] ext4_has_metadata_csum
```

Here are the commands to reproduce:
```
sudo perf kvm --host --guest --guestkallsyms=guest-kallsyms --guestmodules=guest-modules record -p 5175 -o perf.data sleep 10
sudo perf kvm --guest --guestmodules=guest-modules --guestkallsyms=guest-kallsyms report -i perf.data --force 
```

I am going to continue playing with the output. If I find any problems, I will report it but I think now we can close the issue. Thank you for your help!
